# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [1]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    #special_words = ['<PAD>', '<UNK>', '<GO>',  '<EOS>']
    
    source_id_text = [[source_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in line.split()] for line in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int.get(word, target_vocab_to_int['<UNK>']) for word in line.split()] + [target_vocab_to_int['<EOS>']] for line in target_text.split('\n')] 
    
    return (source_id_text, target_id_text)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

NameError: name 'tests' is not defined

### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [9]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [3]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [4]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_data = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    lr = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')

    target_sequence_length = tf.placeholder(tf.int32, (None,), name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, (None,), name='source_sequence_length')
    
    
    return input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [7]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(enc_cell, output_keep_prob=keep_prob)
        return drop

    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [21]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # Training Decoder

    # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                        sequence_length=target_sequence_length,
                                                        time_major=False)

    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                       training_helper,
                                                       encoder_state,
                                                       output_layer) 

    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                impute_finished=True,
                                                                maximum_iterations=max_summary_length)[0]    
    
    return training_decoder_output

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [22]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    *:param encoder_state: Encoder state
    *:param dec_cell: Decoder RNN Cell
    *:param dec_embeddings: Decoder embeddings
    *:param start_of_sequence_id: GO ID
    *:param end_of_sequence_id: EOS Id
    *:param max_target_sequence_length: Maximum length of target sequences
:param vocab_size: Size of decoder/target vocabulary
NOT IN METHOD ASSIGNATURE*:param decoding_scope: TenorFlow Variable Scope for decoding
    *:param output_layer: Function to apply the output layer
    *:param batch_size: Batch size
:param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    #with tf.variable_scope("decode", reuse=True):
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Helper for the inference process.
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                                start_tokens,
                                                                end_of_sequence_id)

    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        inference_helper,
                                                        encoder_state,
                                                        output_layer)

    # Perform dynamic decoding using the decoder
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                        impute_finished=True,
                                                        maximum_iterations=max_target_sequence_length)[0]


    
    return inference_decoder_output
    #return training_decoder_output, inference_decoder_output
    



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [33]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Embeddings
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

     # Decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        
        ## SHOULD I add DROPOUT????
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    
    with tf.variable_scope("decode"):
        # Get the training logits.        
        dec_train_layer_out = decoding_layer_train(encoder_state, 
                                                   dec_cell, 
                                                   dec_embed_input, 
                                                   target_sequence_length,
                                                   max_target_sequence_length, 
                                                   output_layer, 
                                                   keep_prob) 
    with tf.variable_scope("decode",reuse=True):        
        # Get the inference logits.    
        dec_infer_layer_out = decoding_layer_infer(encoder_state, 
                                                  dec_cell, 
                                                  dec_embeddings, 
                                                  target_vocab_to_int['<GO>'], 
                                                  target_vocab_to_int['<EOS>'], 
                                                  max_target_sequence_length, 
                                                  target_vocab_size, #??? not used inside the function
                                                  output_layer, 
                                                  batch_size, 
                                                  keep_prob)
        
    return dec_train_layer_out, dec_infer_layer_out



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [34]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,##parameter not described in the comments below
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Encode the input 
    encoded_output, encoded_state = encoding_layer(input_data, #rnn_inputs
                                                   rnn_size, 
                                                   num_layers, 
                                                   keep_prob,  
                                                   source_sequence_length, 
                                                   source_vocab_size,
                                                   enc_embedding_size)
    
    # Process target data 
    decoder_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    #Decode the encoded input 
    train_decoder_output, inf_decoder_output = decoding_layer(decoder_input, 
                                                              encoded_state, 
                                                              target_sequence_length, 
                                                              max_target_sentence_length, 
                                                              rnn_size, 
                                                              num_layers, 
                                                              target_vocab_to_int, 
                                                              target_vocab_size, 
                                                              batch_size, 
                                                              keep_prob, 
                                                              dec_embedding_size)
    return (train_decoder_output, inf_decoder_output)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [48]:
# Number of Epochs
epochs = 100
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 64
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 100
decoding_embedding_size = 100
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.8
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [49]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [50]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [51]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   50/538 - Train Accuracy: 0.4051, Validation Accuracy: 0.4490, Loss: 2.6275
Epoch   0 Batch  100/538 - Train Accuracy: 0.4807, Validation Accuracy: 0.5067, Loss: 2.0350
Epoch   0 Batch  150/538 - Train Accuracy: 0.4900, Validation Accuracy: 0.5202, Loss: 1.6632
Epoch   0 Batch  200/538 - Train Accuracy: 0.4574, Validation Accuracy: 0.5048, Loss: 1.2201
Epoch   0 Batch  250/538 - Train Accuracy: 0.5404, Validation Accuracy: 0.5664, Loss: 0.9415
Epoch   0 Batch  300/538 - Train Accuracy: 0.5858, Validation Accuracy: 0.5881, Loss: 0.7457
Epoch   0 Batch  350/538 - Train Accuracy: 0.5893, Validation Accuracy: 0.6010, Loss: 0.6858
Epoch   0 Batch  400/538 - Train Accuracy: 0.6179, Validation Accuracy: 0.6168, Loss: 0.6010
Epoch   0 Batch  450/538 - Train Accuracy: 0.6417, Validation Accuracy: 0.6404, Loss: 0.5515
Epoch   0 Batch  500/538 - Train Accuracy: 0.6880, Validation Accuracy: 0.6788, Loss: 0.4677
Epoch   1 Batch   50/538 - Train Accuracy: 0.6820, Validation Accuracy

Epoch   8 Batch  500/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9515, Loss: 0.0197
Epoch   9 Batch   50/538 - Train Accuracy: 0.9338, Validation Accuracy: 0.9501, Loss: 0.0300
Epoch   9 Batch  100/538 - Train Accuracy: 0.9789, Validation Accuracy: 0.9592, Loss: 0.0243
Epoch   9 Batch  150/538 - Train Accuracy: 0.9631, Validation Accuracy: 0.9471, Loss: 0.0262
Epoch   9 Batch  200/538 - Train Accuracy: 0.9666, Validation Accuracy: 0.9569, Loss: 0.0239
Epoch   9 Batch  250/538 - Train Accuracy: 0.9631, Validation Accuracy: 0.9400, Loss: 0.0293
Epoch   9 Batch  300/538 - Train Accuracy: 0.9412, Validation Accuracy: 0.9325, Loss: 0.0331
Epoch   9 Batch  350/538 - Train Accuracy: 0.9550, Validation Accuracy: 0.9503, Loss: 0.0316
Epoch   9 Batch  400/538 - Train Accuracy: 0.9587, Validation Accuracy: 0.9588, Loss: 0.0280
Epoch   9 Batch  450/538 - Train Accuracy: 0.9163, Validation Accuracy: 0.9402, Loss: 0.0464
Epoch   9 Batch  500/538 - Train Accuracy: 0.9686, Validation Accuracy

Epoch  17 Batch  450/538 - Train Accuracy: 0.9280, Validation Accuracy: 0.9659, Loss: 0.0402
Epoch  17 Batch  500/538 - Train Accuracy: 0.9743, Validation Accuracy: 0.9572, Loss: 0.0205
Epoch  18 Batch   50/538 - Train Accuracy: 0.9441, Validation Accuracy: 0.9501, Loss: 0.0271
Epoch  18 Batch  100/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9547, Loss: 0.0225
Epoch  18 Batch  150/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9426, Loss: 0.0217
Epoch  18 Batch  200/538 - Train Accuracy: 0.9781, Validation Accuracy: 0.9613, Loss: 0.0159
Epoch  18 Batch  250/538 - Train Accuracy: 0.9666, Validation Accuracy: 0.9625, Loss: 0.0227
Epoch  18 Batch  300/538 - Train Accuracy: 0.9701, Validation Accuracy: 0.9467, Loss: 0.0237
Epoch  18 Batch  350/538 - Train Accuracy: 0.9444, Validation Accuracy: 0.9421, Loss: 0.0282
Epoch  18 Batch  400/538 - Train Accuracy: 0.9712, Validation Accuracy: 0.9522, Loss: 0.0246
Epoch  18 Batch  450/538 - Train Accuracy: 0.9448, Validation Accuracy

Epoch  26 Batch  400/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9423, Loss: 0.0269
Epoch  26 Batch  450/538 - Train Accuracy: 0.9412, Validation Accuracy: 0.9434, Loss: 0.0359
Epoch  26 Batch  500/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9574, Loss: 0.0175
Epoch  27 Batch   50/538 - Train Accuracy: 0.9590, Validation Accuracy: 0.9634, Loss: 0.0222
Epoch  27 Batch  100/538 - Train Accuracy: 0.9779, Validation Accuracy: 0.9602, Loss: 0.0187
Epoch  27 Batch  150/538 - Train Accuracy: 0.9715, Validation Accuracy: 0.9592, Loss: 0.0212
Epoch  27 Batch  200/538 - Train Accuracy: 0.9707, Validation Accuracy: 0.9593, Loss: 0.0164
Epoch  27 Batch  250/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9524, Loss: 0.0234
Epoch  27 Batch  300/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9569, Loss: 0.0207
Epoch  27 Batch  350/538 - Train Accuracy: 0.9578, Validation Accuracy: 0.9579, Loss: 0.0321
Epoch  27 Batch  400/538 - Train Accuracy: 0.9730, Validation Accuracy

Epoch  35 Batch  350/538 - Train Accuracy: 0.9647, Validation Accuracy: 0.9609, Loss: 0.0232
Epoch  35 Batch  400/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9652, Loss: 0.0151
Epoch  35 Batch  450/538 - Train Accuracy: 0.9472, Validation Accuracy: 0.9517, Loss: 0.0350
Epoch  35 Batch  500/538 - Train Accuracy: 0.9831, Validation Accuracy: 0.9608, Loss: 0.0123
Epoch  36 Batch   50/538 - Train Accuracy: 0.9564, Validation Accuracy: 0.9558, Loss: 0.0255
Epoch  36 Batch  100/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9512, Loss: 0.0211
Epoch  36 Batch  150/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9473, Loss: 0.0171
Epoch  36 Batch  200/538 - Train Accuracy: 0.9789, Validation Accuracy: 0.9622, Loss: 0.0137
Epoch  36 Batch  250/538 - Train Accuracy: 0.9648, Validation Accuracy: 0.9560, Loss: 0.0205
Epoch  36 Batch  300/538 - Train Accuracy: 0.9673, Validation Accuracy: 0.9657, Loss: 0.0160
Epoch  36 Batch  350/538 - Train Accuracy: 0.9552, Validation Accuracy

Epoch  44 Batch  300/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9482, Loss: 0.0194
Epoch  44 Batch  350/538 - Train Accuracy: 0.9548, Validation Accuracy: 0.9421, Loss: 0.0270
Epoch  44 Batch  400/538 - Train Accuracy: 0.9697, Validation Accuracy: 0.9586, Loss: 0.0214
Epoch  44 Batch  450/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9549, Loss: 0.0374
Epoch  44 Batch  500/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9586, Loss: 0.0201
Epoch  45 Batch   50/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9664, Loss: 0.0250
Epoch  45 Batch  100/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9588, Loss: 0.0163
Epoch  45 Batch  150/538 - Train Accuracy: 0.9707, Validation Accuracy: 0.9458, Loss: 0.0219
Epoch  45 Batch  200/538 - Train Accuracy: 0.9750, Validation Accuracy: 0.9490, Loss: 0.0184
Epoch  45 Batch  250/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9464, Loss: 0.0277
Epoch  45 Batch  300/538 - Train Accuracy: 0.9667, Validation Accuracy

Epoch  53 Batch  250/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9600, Loss: 0.0308
Epoch  53 Batch  300/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9631, Loss: 0.0256
Epoch  53 Batch  350/538 - Train Accuracy: 0.9529, Validation Accuracy: 0.9620, Loss: 0.0289
Epoch  53 Batch  400/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9501, Loss: 0.0194
Epoch  53 Batch  450/538 - Train Accuracy: 0.9490, Validation Accuracy: 0.9480, Loss: 0.0377
Epoch  53 Batch  500/538 - Train Accuracy: 0.9773, Validation Accuracy: 0.9599, Loss: 0.0216
Epoch  54 Batch   50/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9659, Loss: 0.0242
Epoch  54 Batch  100/538 - Train Accuracy: 0.9729, Validation Accuracy: 0.9705, Loss: 0.0179
Epoch  54 Batch  150/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9604, Loss: 0.0200
Epoch  54 Batch  200/538 - Train Accuracy: 0.9713, Validation Accuracy: 0.9496, Loss: 0.0176
Epoch  54 Batch  250/538 - Train Accuracy: 0.9686, Validation Accuracy

Epoch  62 Batch  200/538 - Train Accuracy: 0.9844, Validation Accuracy: 0.9604, Loss: 0.0158
Epoch  62 Batch  250/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9691, Loss: 0.0220
Epoch  62 Batch  300/538 - Train Accuracy: 0.9596, Validation Accuracy: 0.9544, Loss: 0.0245
Epoch  62 Batch  350/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9512, Loss: 0.0260
Epoch  62 Batch  400/538 - Train Accuracy: 0.9714, Validation Accuracy: 0.9508, Loss: 0.0237
Epoch  62 Batch  450/538 - Train Accuracy: 0.9557, Validation Accuracy: 0.9679, Loss: 0.0371
Epoch  62 Batch  500/538 - Train Accuracy: 0.9748, Validation Accuracy: 0.9556, Loss: 0.0226
Epoch  63 Batch   50/538 - Train Accuracy: 0.9607, Validation Accuracy: 0.9590, Loss: 0.0193
Epoch  63 Batch  100/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9609, Loss: 0.0230
Epoch  63 Batch  150/538 - Train Accuracy: 0.9648, Validation Accuracy: 0.9489, Loss: 0.0219
Epoch  63 Batch  200/538 - Train Accuracy: 0.9762, Validation Accuracy

Epoch  71 Batch  150/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9545, Loss: 0.0171
Epoch  71 Batch  200/538 - Train Accuracy: 0.9842, Validation Accuracy: 0.9608, Loss: 0.0180
Epoch  71 Batch  250/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9522, Loss: 0.0235
Epoch  71 Batch  300/538 - Train Accuracy: 0.9673, Validation Accuracy: 0.9622, Loss: 0.0203
Epoch  71 Batch  350/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9696, Loss: 0.0227
Epoch  71 Batch  400/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9602, Loss: 0.0195
Epoch  71 Batch  450/538 - Train Accuracy: 0.9475, Validation Accuracy: 0.9647, Loss: 0.0352
Epoch  71 Batch  500/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9478, Loss: 0.0154
Epoch  72 Batch   50/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9648, Loss: 0.0199
Epoch  72 Batch  100/538 - Train Accuracy: 0.9779, Validation Accuracy: 0.9652, Loss: 0.0169
Epoch  72 Batch  150/538 - Train Accuracy: 0.9723, Validation Accuracy

Epoch  80 Batch  100/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9712, Loss: 0.0192
Epoch  80 Batch  150/538 - Train Accuracy: 0.9670, Validation Accuracy: 0.9632, Loss: 0.0275
Epoch  80 Batch  200/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9609, Loss: 0.0261
Epoch  80 Batch  250/538 - Train Accuracy: 0.9707, Validation Accuracy: 0.9563, Loss: 0.0226
Epoch  80 Batch  300/538 - Train Accuracy: 0.9697, Validation Accuracy: 0.9560, Loss: 0.0225
Epoch  80 Batch  350/538 - Train Accuracy: 0.9740, Validation Accuracy: 0.9558, Loss: 0.0308
Epoch  80 Batch  400/538 - Train Accuracy: 0.9637, Validation Accuracy: 0.9434, Loss: 0.0203
Epoch  80 Batch  450/538 - Train Accuracy: 0.9589, Validation Accuracy: 0.9560, Loss: 0.0350
Epoch  80 Batch  500/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9528, Loss: 0.0206
Epoch  81 Batch   50/538 - Train Accuracy: 0.9625, Validation Accuracy: 0.9581, Loss: 0.0218
Epoch  81 Batch  100/538 - Train Accuracy: 0.9570, Validation Accuracy

Epoch  89 Batch   50/538 - Train Accuracy: 0.9557, Validation Accuracy: 0.9656, Loss: 0.0233
Epoch  89 Batch  100/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9712, Loss: 0.0165
Epoch  89 Batch  150/538 - Train Accuracy: 0.9643, Validation Accuracy: 0.9593, Loss: 0.0274
Epoch  89 Batch  200/538 - Train Accuracy: 0.9715, Validation Accuracy: 0.9677, Loss: 0.0173
Epoch  89 Batch  250/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9611, Loss: 0.0276
Epoch  89 Batch  300/538 - Train Accuracy: 0.9656, Validation Accuracy: 0.9622, Loss: 0.0253
Epoch  89 Batch  350/538 - Train Accuracy: 0.9578, Validation Accuracy: 0.9625, Loss: 0.0211
Epoch  89 Batch  400/538 - Train Accuracy: 0.9697, Validation Accuracy: 0.9673, Loss: 0.0222
Epoch  89 Batch  450/538 - Train Accuracy: 0.9464, Validation Accuracy: 0.9689, Loss: 0.0447
Epoch  89 Batch  500/538 - Train Accuracy: 0.9663, Validation Accuracy: 0.9457, Loss: 0.0207
Epoch  90 Batch   50/538 - Train Accuracy: 0.9586, Validation Accuracy

Epoch  97 Batch  500/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9553, Loss: 0.0215
Epoch  98 Batch   50/538 - Train Accuracy: 0.9633, Validation Accuracy: 0.9657, Loss: 0.0254
Epoch  98 Batch  100/538 - Train Accuracy: 0.9793, Validation Accuracy: 0.9680, Loss: 0.0172
Epoch  98 Batch  150/538 - Train Accuracy: 0.9605, Validation Accuracy: 0.9648, Loss: 0.0232
Epoch  98 Batch  200/538 - Train Accuracy: 0.9707, Validation Accuracy: 0.9654, Loss: 0.0214
Epoch  98 Batch  250/538 - Train Accuracy: 0.9662, Validation Accuracy: 0.9634, Loss: 0.0190
Epoch  98 Batch  300/538 - Train Accuracy: 0.9567, Validation Accuracy: 0.9467, Loss: 0.0210
Epoch  98 Batch  350/538 - Train Accuracy: 0.9712, Validation Accuracy: 0.9611, Loss: 0.0301
Epoch  98 Batch  400/538 - Train Accuracy: 0.9621, Validation Accuracy: 0.9677, Loss: 0.0218
Epoch  98 Batch  450/538 - Train Accuracy: 0.9440, Validation Accuracy: 0.9457, Loss: 0.0356
Epoch  98 Batch  500/538 - Train Accuracy: 0.9714, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [52]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [53]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [61]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    return [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split()]

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [71]:
translate_sentence = 'he saw a old yellow truck .'
#translate_sentence = 'this fruit is an orange . '


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [84, 157, 212, 20, 184, 228, 208]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [127, 131, 107, 179, 311, 349, 161, 1]
  French Words: il conduit une vieille voiture jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.